group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -5524,5 +5670,2897 @@ private static Record newGroupMetadataRecord(
+        GroupMetadataManagerTestContext context = new GroupMetadataManagerTestContext.Builder()
+            .build();
+        GenericGroup group = context.createGenericGroup("group-id");
+
+        // JoinGroup(leader) with the Protocol Type of the group
+        JoinGroupRequestData joinRequest = new JoinGroupRequestBuilder()
+            .withGroupId("group-id")
+            .withGroupInstanceId("leader-instance-id")
+            .withMemberId(UNKNOWN_MEMBER_ID)
+            .withProtocolSuperset()
+            .build();
+
+        CompletableFuture<JoinGroupResponseData> leaderJoinFuture = new CompletableFuture<>();
+        CoordinatorResult<Void, Record> result = context.sendGenericGroupJoin(joinRequest, leaderJoinFuture);
+
+        assertTrue(result.records().isEmpty());
+        assertFalse(leaderJoinFuture.isDone());
+
+        // JoinGroup(follower) with the Protocol Type of the group
+        CompletableFuture<JoinGroupResponseData> followerJoinFuture = new CompletableFuture<>();
+        result = context.sendGenericGroupJoin(joinRequest.setGroupInstanceId("follower-instance-id"), followerJoinFuture);
+
+        assertTrue(result.records().isEmpty());
+        assertFalse(followerJoinFuture.isDone());
+
+        // Advance clock by rebalance timeout to complete join phase.
+        assertNoOrEmptyResult(context.sleep(10000));
+        assertTrue(leaderJoinFuture.isDone());
+        assertTrue(followerJoinFuture.isDone());
+        assertEquals(Errors.NONE.code(), leaderJoinFuture.get().errorCode());
+        assertEquals(Errors.NONE.code(), followerJoinFuture.get().errorCode());
+
+        String leaderId = leaderJoinFuture.get().memberId();
+        String followerId = followerJoinFuture.get().memberId();
+        int generationId = leaderJoinFuture.get().generationId();
+
+        // SyncGroup with the provided Protocol Type and Name
+        List<SyncGroupRequestAssignment> assignment = new ArrayList<>();
+        assignment.add(new SyncGroupRequestAssignment().setMemberId(leaderId));
+        SyncGroupRequestData syncRequest = new SyncGroupRequestBuilder()
+            .withGroupId("group-id")
+            .withMemberId(leaderId)
+            .withProtocolType(protocolType.orElse(null))
+            .withProtocolName(protocolName.orElse(null))
+            .withGenerationId(generationId)
+            .withAssignment(assignment)
+            .build();
+
+        CompletableFuture<SyncGroupResponseData> leaderSyncFuture = new CompletableFuture<>();
+        result = context.sendGenericGroupSync(syncRequest, leaderSyncFuture);
+
+        if (expectedError == Errors.NONE) {
+            assertEquals(
+                Collections.singletonList(newGroupMetadataRecord(group, MetadataVersion.latest())),
+                result.records()
+            );
+            // Simulate successful write to log.
+            result.appendFuture().complete(null);
+        } else {
+            assertTrue(result.records().isEmpty());
+        }
```

Comment:
In my opinion, we don't have to validate the records in all the tests. It is much better to simplify most of tests and to have a strong set of tests to verify them. In this particular case, it does not bring any value as we want to test the protocol type/name consistency.

Reply:
makes sense, i removed it for most of the tests, at least for the ones that generate records